### PR TITLE
baseline: Use constant gas check if possible

### PR DESCRIPTION
**Pull request by chfast**
If instruction base gas cost is constant across EVM revisions, use the inlined constant in the code directly (no table lookup).
Also change the order of requirements checks: check the stack first. This also helps performance.

### CODE REVIEW

**Comment by chfast** [comment]
Benchmarks look good.
```
baseline/execute/main/blake2b_huff/empty_mean                               -0.0685         -0.0685            14            13            14            13
baseline/execute/main/blake2b_huff/2805nulls_mean                           -0.0649         -0.0649           302           283           302           283
baseline/execute/main/blake2b_huff/5610nulls_mean                           -0.0680         -0.0680           590           550           590           550
baseline/execute/main/blake2b_huff/8415nulls_mean                           -0.0652         -0.0652           865           808           865           808
baseline/execute/main/blake2b_huff/65536nulls_mean                          -0.0677         -0.0677          6725          6270          6725          6270
baseline/execute/main/blake2b_shifts/2805nulls_mean                         -0.0408         -0.0408          2915          2796          2915          2796
baseline/execute/main/blake2b_shifts/5610nulls_mean                         -0.0406         -0.0406          5807          5571          5807          5571
baseline/execute/main/blake2b_shifts/8415nulls_mean                         -0.0423         -0.0423          8708          8339          8708          8339
baseline/execute/main/blake2b_shifts/65536nulls_mean                        -0.0424         -0.0424         67484         64624         67485         64624
baseline/execute/main/sha1_divs/empty_mean                                  -0.0261         -0.0261            52            51            52            51
baseline/execute/main/sha1_divs/1351_mean                                   -0.0257         -0.0257          1094          1065          1094          1065
baseline/execute/main/sha1_divs/2737_mean                                   -0.0255         -0.0255          2134          2080          2134          2080
baseline/execute/main/sha1_divs/5311_mean                                   -0.0256         -0.0256          4169          4062          4169          4062
baseline/execute/main/sha1_divs/65536_mean                                  -0.0244         -0.0244         50783         49544         50783         49544
baseline/execute/main/sha1_shifts/empty_mean                                -0.0418         -0.0418            31            30            31            30
baseline/execute/main/sha1_shifts/1351_mean                                 -0.0433         -0.0433           666           638           666           638
baseline/execute/main/sha1_shifts/2737_mean                                 -0.0435         -0.0435          1301          1244          1301          1244
baseline/execute/main/sha1_shifts/5311_mean                                 -0.0440         -0.0440          2540          2429          2540          2429
baseline/execute/main/sha1_shifts/65536_mean                                -0.0448         -0.0448         30965         29579         30965         29579
baseline/execute/main/weierstrudel/0_mean                                   +0.0025         +0.0025           157           157           157           157
baseline/execute/main/weierstrudel/1_mean                                   -0.0088         -0.0088           347           344           347           344
baseline/execute/main/weierstrudel/3_mean                                   -0.0120         -0.0120           546           539           546           539
baseline/execute/main/weierstrudel/9_mean                                   -0.0164         -0.0164          1134          1115          1134          1115
baseline/execute/main/weierstrudel/14_mean                                  -0.0173         -0.0173          1626          1598          1626          1598
baseline/execute/micro/jump_around/empty_mean                               -0.1068         -0.1068            24            22            24            22
baseline/execute/micro/jumpdests_0xffff/empty_mean                          -0.0584         -0.0584           113           107           113           107
baseline/execute/micro/loop_with_many_jumpdests/empty_mean                  -0.0369         -0.0369         13149         12663         13149         12663
baseline/execute/micro/memory_grow_mload/nogrow_mean                        -0.0035         -0.0035            51            51            51            51
baseline/execute/micro/memory_grow_mload/by1_mean                           +0.0000         +0.0000            53            53            53            53
baseline/execute/micro/memory_grow_mload/by16_mean                          -0.0258         -0.0258            60            58            60            58
baseline/execute/micro/memory_grow_mload/by32_mean                          -0.0360         -0.0360            69            66            69            66
baseline/execute/micro/memory_grow_mstore/nogrow_mean                       -0.0153         -0.0153            50            50            50            50
baseline/execute/micro/memory_grow_mstore/by1_mean                          -0.0088         -0.0088            54            53            54            53
baseline/execute/micro/memory_grow_mstore/by16_mean                         -0.0300         -0.0300            61            59            61            59
baseline/execute/micro/memory_grow_mstore/by32_mean                         -0.0467         -0.0467            71            68            71            68
baseline/execute/micro/signextend/zero_mean                                 -0.0824         -0.0824            69            63            69            63
baseline/execute/micro/signextend/one_mean                                  -0.0839         -0.0839            69            63            69            63
```

File: lib/evmone/instruction_traits.hpp [hunk 1]
```diff
@@ -184,6 +184,27 @@ struct Traits
     std::optional<evmc_revision> since;
 };
 
+/// Determines if an instruction has constant base gas cost across or revisions.
```
Comment (gumb0) [inline lib/evmone/instruction_traits.hpp:187]:
```suggestion
/// Determines if an instruction has constant base gas cost across all revisions.
```

File: lib/evmone/baseline.cpp [hunk 2]
```diff
@@ -51,12 +51,19 @@ template <evmc_opcode Op>
...
+    auto gas_cost = instr::gas_costs[EVMC_FRONTIER][Op];  // Init assuming const cost.
+    if constexpr (!instr::has_const_gas_cost(Op))
+    {
+        gas_cost = cost_table[Op];  // If not, load the cost from the table.
 
-    if (INTX_UNLIKELY((state.gas_left -= cost_table[Op]) < 0))
-        return EVMC_OUT_OF_GAS;
+        // Negative cost marks an undefined instruction.
+        // This check must be first to produce correct error code.
+        if (INTX_UNLIKELY(gas_cost < 0))
```
Comment (gumb0) [inline lib/evmone/baseline.cpp:61]: Does the change from `== instr::undefined` improves performance?

Comment (gumb0) [reply lib/evmone/baseline.cpp:61]: As it's now done only for non-const-gas instructions, this function will not produce correct result for always undefined instructions, so add something like
```
static_assert(instr::gas_costs[EVMC_FRONTIER][Op] != instr::undefined || !instr::has_const_gas_cost(Op));
```

**Review by gumb0 — APPROVED** [review]

File: lib/evmone/baseline.cpp (hunk 2 above)
Comment (chfast) [reply lib/evmone/baseline.cpp:61]: > Does the change from `== instr::undefined` improves performance?

Not really. It saves one byte of the code.

**Pull request merged**